libraries/AP_Baro/examples/ICM20789/ICM20789.cpp

```diff
@@ -0,0 +1,291 @@
+
+    // dump registers
+    printf("ICM20789 registers\n");
+#if 0
+    for (uint8_t reg = 0; reg<0x80; reg++) {
+        v=0;
+        spi_dev->read_registers(reg, &v, 1);
+        printf("%02x:%02x ", (unsigned)reg, (unsigned)v);
+        if ((reg+1) % 16 == 0) {
+            printf("\n");
+        }
+    }
+    printf("\n");
+#endif
+    
+    spi_dev->get_semaphore()->give();
+}
+
+
+/*
+  send a 16 bit command to the baro
+ */
+static bool send_cmd16(uint16_t cmd)
+{
+    uint8_t cmd_b[2] = { uint8_t(cmd >> 8), uint8_t(cmd & 0xFF) };
+    return i2c_dev->transfer(cmd_b, 2, nullptr, 0);
+}
+
+/*
+  read baro calibration data
+ */
+static bool read_calibration_data(void)
+{
+    // setup for OTP read
+    const uint8_t cmd[5] = { 0xC5, 0x95, 0x00, 0x66, 0x9C };
+    if (!i2c_dev->transfer(cmd, sizeof(cmd), nullptr, 0)) {
+        return false;
+    }
+    for (uint8_t i=0; i<4; i++) {
+        if (!send_cmd16(0xC7F7)) {
+            return false;
+        }
+        uint8_t d[3];
+        if (!i2c_dev->transfer(nullptr, 0, d, sizeof(d))) {
+            return false;
+        }
+        uint16_t c = int16_t((d[0]<<8) | d[1]);
+        printf("sensor_constants[%u]=%d\n", i, c);
+    }
+    return true;
+}
+
+// initialise baro on i2c
+static void i2c_init(void)
+{
+    if (!i2c_dev->get_semaphore()->take(0)) {
+        AP_HAL::panic("Failed to get baro semaphore");
+    }
+       
+    if (send_cmd16(CMD_READ_ID)) {
+        printf("ICM20789: read ID ******\n");
+        uint8_t id[3] {};
+        if (!i2c_dev->transfer(nullptr, 0, id, 3)) {
+            printf("ICM20789: failed to read ID\n");        
+        }
+        printf("ICM20789: ID %02x %02x %02x\n", id[0], id[1], id[2]);
+    } else {
+        printf("ICM20789 read ID failed\n");
+    }
+
+    if (send_cmd16(CMD_SOFT_RESET)) {
+        printf("ICM20789: reset OK ************###########*********!!!!!!!!\n");
+    } else {
+        printf("ICM20789 baro reset failed\n");
+    }
+    hal.scheduler->delay(1);
+        
+    
+    read_calibration_data();
+
+    i2c_dev->get_semaphore()->give();
+
+    printf("checking lidar\n");
+    if (!lidar_dev->get_semaphore()->take(0)) {
+        AP_HAL::panic("Failed to get lidar semaphore");
+    }
+
+    uint8_t regs[3] = { 0xC0, 0xC1, 0xC2 };
+    for (uint8_t i=0; i<ARRAY_SIZE(regs); i++) {
+        uint8_t v = 0;
+        lidar_dev->read_registers(regs[i], &v, 1);
+        printf("0x%02x : 0x%02x\n", regs[i], v);
+    }
+    lidar_dev->get_semaphore()->give();    
+}
+
+void setup()
+{
+    printf("ICM20789 test\n");
+
+    AP_BoardConfig{}.init();
+
+    hal.scheduler->delay(1000);
+
+#ifdef HAL_INS_MPU60x0_NAME
+    spi_dev = std::move(hal.spi->get_device(HAL_INS_MPU60x0_NAME));
+
+    if (!spi_dev->get_semaphore()->take(0)) {
+        AP_HAL::panic("Failed to get spi semaphore");
+    }
+
+
+    i2c_dev = std::move(hal.i2c_mgr->get_device(1, 0x63));
+    lidar_dev = std::move(hal.i2c_mgr->get_device(1, 0x29));
```

Comment:
What is the VL53L0X doing in the middle of this example? :stuck_out_tongue_winking_eye:

Reply:
Will generally fix the example.